src/basicfuns.jl

```diff
@@ -152,9 +152,43 @@ Return `log(1+exp(x))` evaluated carefully for largish `x`.
 
 This is also called the ["softplus"](https://en.wikipedia.org/wiki/Rectifier_(neural_networks))
 transformation, being a smooth approximation to `max(0,x)`. Its inverse is [`logexpm1`](@ref).
+
+See:
+ * Martin Maechler (2012) [“Accurately Computing log(1 − exp(− |a|))”](http://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf)
+
+Note: different than Maechler (2012), also uses bounds specific to Float32 and Float16.
 """
-log1pexp(x::Real) = x < 18.0 ? log1p(exp(x)) : x < 33.3 ? x + exp(-x) : oftype(exp(-x), x)
-log1pexp(x::Float32) = x < 9.0f0 ? log1p(exp(x)) : x < 16.0f0 ? x + exp(-x) : oftype(exp(-x), x)
+function log1pexp(x::Real)
+    t = log1p(exp(-abs(x)))
+    return x ≤ 0 ? t : t + x
+end
+
+log1pexp(x::Union{Float16,Float32,Float64}) = _log1pexp_approx(x)
+
+function _log1pexp_approx(x::Real)
+	a, b, c = _log1pexp_branch_bounds(x)
+    if x ≤ a
+        return exp(x)
+    elseif x ≤ b
+        return log1p(exp(x))
+    elseif x ≤ c
+        return x + exp(-x)
+    else
+        return x
+    end
+end
+
+#=
+Given the `approx` used in a branch of log1pexp(x) above, we find the first `x` (from above
+or below) that is a root of
+
+    T(log1pexp(big(x))) - approx(T(x))
+
+This determines the branch bounds below.
+=#
+_log1pexp_branch_bounds(::Float64) = (-37.0, 18.0, 33.3)
+_log1pexp_branch_bounds(::Float32) = (-15f0, 9f0, 14.5f0)
+_log1pexp_branch_bounds(::Float16) = (Float16(-7), Float16(3), Float16(5.7))
```

Comment:
This seems a bit complicated, in particular I don't think we should introduce so many levels of functions. Maybe just use

```suggestion
# The thresholds below are determined by the first root `x` (from above or below)
#of `T(log1pexp(big(x))) - approx(T(x))``, where `approx` is the approximation of interest.
for (T, a, b, c) in ((:Float64, -37.0, 18.0, 33.3), (:Float32, -15f0, 9f0, 14.5f0), (:Float16, Float16(-7), Float16(3), Float16(5.7))
    @eval function function log1pexp(x::$T)
        if x ≤ $a
            return exp(x)
        elseif x ≤ $b
            return log1p(exp(x))
        elseif x ≤ $c
            return x + exp(-x)
        else
            return x
        end
    end
end
```

Reply:
Hmm I'd prefer to go back to

```
function log1pexp(x::Union{Float16,Float32,Float64})
	a, b, c = _log1pexp_branch_bounds(x)
    if x ≤ a
        return exp(x)
    elseif x ≤ b
        return log1p(exp(x))
    elseif x ≤ c
        return x + exp(-x)
    else
        return x
    end
end
```

It's the same pattern used by `logistic`.

Reply:
I still don't understand how what you propose with `@eval` is more "extensible" than this ...

Reply:
No, of course, it's not more extendable. The point is that it avoids defining any other function besides `log1pexp` since these can't be extended either in the version you suggest.